### PR TITLE
Set peer before sending request

### PR DESCRIPTION
**Pull request by jack0**
If a connection to the server exists but a request errors out due to context canceled or deadline exceeded, the peer information is not set. We can move setting the peer out of the recvResponse and before the sendRequest to get peer information for some failed RPCs. This addresses #1417

@MakMukhi @therc

### CODE REVIEW

**Comment by markdroth** [comment]
It's worth noting that this may impact the retry code that @dfawley is working on (and I'm running into a similar problem in C-core).  The peer that we are talking to when we send the first message may not be the one that ultimately successfully handles the request, so if we set the peer when we send the first message and then retry on a different backend later, the peer information we originally set may be incorrect.  We probably need to discuss the right semantics for returning peer information in the context of retries (and the hedging case is even more unclear, since we may be talking to multiple backends simultaneously).

**Comment by therc** [comment]
Retries are the kind of scenario I had in mind when I said in the original issue that maybe the peer could be set twice. And yeah, hedging is even trickier. Personally, I'd be OK if initially gRPC just tracked the last peer that succeeded/failed.
Our use case: we're trying to debug an issue with canceled contexts that does not manifest itself in a more controlled environment.

**Comment by dfawley** [comment]
@markdroth interesting, thanks for calling this to my attention.  I think this change is OK for us as long as nobody accesses the peer while the RPC is happening.  Our API is synchronous, so this would be unexpected.  For retries, it would end up being set to the _last_ attempt, which is presumably what we want.  For hedging, we will have to move things around a bit, but this won't make things much worse than they already are in that regard.

File: test/end2end_test.go [hunk 1]
```diff
@@ -2263,6 +2263,55 @@ func testPeerNegative(t *testing.T, e env) {
...
+			_, sp, err := net.SplitHostPort(te.srvAddr)
+			if err != nil {
+				t.Fatalf("Failed to parse address of test server.")
+			}
+			if pp != sp {
+				t.Fatalf("peer.Addr = localhost:%v, want localhost:%v", pp, sp)
+			}
+			break
+		}
+		time.Sleep(10 * time.Millisecond)
```
Comment (MakMukhi) [inline test/end2end_test.go]: Why is the sleep needed?

Comment (jack0) [inline test/end2end_test.go]: It's not needed, I'll remove it.

Comment (jack0) [inline test/end2end_test.go]: I remembered why you need it, you need to give it sufficient time to reach the block of code that sets the peer.

Comment (MakMukhi) [inline test/end2end_test.go]: How about we relied on something else to cause the RPC to fail, like may be trying to send a message size bigger than the max allowed message size?
 This [`call option`](https://github.com/grpc/grpc-go/blob/master/rpc_util.go#L215) can be used to set a max message size allowed for a call. And the test can then deliberately try and send a message bigger than what you set.

Comment (jack0) [inline test/end2end_test.go]: I've made the requested change, thanks!

**Pull request merged**